lucet-idl/src/witx/toplevel.rs

```diff
@@ -0,0 +1,250 @@
+use super::parser::{DeclSyntax, ParseError, TopLevelSyntax};
+use super::sexpr::SExprParser;
+use super::WitxError;
+use std::fs;
+use std::path::{Path, PathBuf};
+
+trait WitxIo {
+    fn fgets(&self, path: &Path) -> Result<String, WitxError>;
+    fn canonicalize(&self, path: &Path) -> Result<PathBuf, WitxError>;
+}
+
+struct Filesystem;
+
+impl WitxIo for Filesystem {
+    fn fgets(&self, path: &Path) -> Result<String, WitxError> {
+        fs::read_to_string(path).map_err(|e| WitxError::UseResolution(path.to_path_buf(), e))
+    }
+    fn canonicalize(&self, path: &Path) -> Result<PathBuf, WitxError> {
+        path.canonicalize()
+            .map_err(|e| WitxError::UseResolution(path.to_path_buf(), e))
+    }
+}
+
+pub fn parse_witx<P: AsRef<Path>>(i: P) -> Result<Vec<DeclSyntax>, WitxError> {
+    parse_witx_with(i, &Filesystem)
+}
+
+fn parse_witx_with<P: AsRef<Path>>(
+    i: P,
+    witxio: &dyn WitxIo,
+) -> Result<Vec<DeclSyntax>, WitxError> {
+    let i_buf = PathBuf::from(i.as_ref());
+    let input_path = witxio.canonicalize(&i_buf)?;
```

Comment:
Looking at this again, I also realized this can be:
```rust
    let input_path = witxio.canonicalize(i.as_ref())?;
```

Reply:
got it